Standardized clamp kernels to Numpy-like implementation

vsimkus opened the pull request:
**BC-breaking note**

For ease of exposition let a_min be the value of the "min" argument to clamp, and a_max be the value of the "max" argument to clamp.

This PR changes the behavior of torch.clamp to always compute min(max(a, a_min), a_max). torch.clamp currently computes this in its vectorized CPU specializations:

https://github.com/pytorch/pytorch/blob/78b95b6204809822def6dd1b06d03cf002cd30c5/aten/src/ATen/cpu/vec256/vec256_double.h#L304

but in other places it clamps differently:

https://github.com/pytorch/pytorch/blob/78b95b6204809822def6dd1b06d03cf002cd30c5/aten/src/ATen/cpu/vec256/vec256_base.h#L624

https://github.com/pytorch/pytorch/blob/78b95b6204809822def6dd1b06d03cf002cd30c5/aten/src/ATen/native/cuda/UnaryOpsKernel.cu#L160

These implementations are the same when a_min < a_max, but divergent when a_min > a_max. This divergence is easily triggered:

```
t = torch.arange(200).to(torch.float)
torch.clamp(t, 4, 2)[0]
: tensor(2.)

torch.clamp(t.cuda(), 4, 2)[0]
: tensor(4., device='cuda:0')

torch.clamp(torch.tensor(0), 4, 2)
: tensor(4)
```

This PR makes the behavior consistent with NumPy's clip. C++'s std::clamp's behavior is undefined when a_min > a_max, but Clang's std::clamp will return 10 in this case (although the program, per the above comment, is in error). Python has no standard clamp implementation.

**PR Summary**

Fixes discrepancy between AVX, CUDA, and base vector implementation for clamp, such that all implementations are consistent and use min(max_vec, max(min_vec, x) formula, thus making it equivalent to numpy.clip in all implementations.

The same fix as in #32587 but isolated to the kernel change only, so that the internal team can benchmark.

mruberry commented:
I ran this through our perf experiments and it passed. So that's good news because we can safely make this change, but bad news in that we don't know what was causing the performance degradation on the larger PR. There's a small chance it was benchmark flakiness, but I would like to propose two options for you, @vsimkus:

- We could update this PR with a test for its behavior and land, then continue to refactor the larger PR into pieces we can validate incrementally with perf experiments
- We could rebase the larger PR and run some clamp-specific benchmarks on it to verify it really does cause a performance regression, hoping the initial experiment failures were due to flakiness

I kind of prefer the first option, myself, since the original PR is now in need of a rebase. What if we separated the changes into this PR, the non-quantization changes, and the quantization changes? 

Looking forward to hearing your thoughts.

vsimkus commented:
@mruberry Thanks! I agree with you - it will be easier to land smaller PRs. I'll be away now for a couple of weeks, but once I return I'll add some tests to this PR and update the docstrings where necessary. Then, we can work on the other bits.

mruberry commented:
> @mruberry Thanks! I agree with you - it will be easier to land smaller PRs. I'll be away now for a couple of weeks, but once I return I'll add some tests to this PR and update the docstrings where necessary. Then, we can work on the other bits.

Great! Looking forward to it.

vsimkus commented:
@mruberry As before, I've refactored the clamp tests to be more concise and use NumPy as the reference implementation. Also updated the docstring so that it shows the correct clamping output formula and removed an unnecessary comment about the argument types. 

I think it should be good for a review and benchmarking now.

mruberry commented on test/test_torch.py:
```diff
@@ -57,6 +57,15 @@
 
 SIZE = 100
 
+def rgetattr(obj, attr, *args):
```
We should get rid of this

mruberry commented on test/test_torch.py:
```diff
@@ -6433,85 +6442,104 @@ def test_logical_and(self, device):
 
-            op(m1, min=min_val, out=out)
-            self.assertEqual(out, res1)
+                op = rgetattr(torch, op_name)
```
Instead of doing this you can just put the ops in the op list:

```
op_list = (torch.clamp, torch.Tensor.clamp, torch.clip, torch.Tensor.clip)
```

mruberry commented on test/test_torch.py:
```diff
@@ -6433,85 +6442,104 @@ def test_logical_and(self, device):
+                Y_actual = op(X, min_val, max_val)
+                self.assertEqual(Y_expected, Y_actual)
+
+                # Test op-out behavior
```
For the out behavior you can just test if `op == torch.clamp`, for example, since that has an out variant but the method versions of the op don't

mruberry commented on test/test_torch.py:
```diff
@@ -6433,85 +6442,104 @@ def test_logical_and(self, device):
+                self.assertEqual(Y_expected, Y_out)
+
+                # Test op-inplace behaviour
+                op_inplace(X, min_val, max_val)
```
Per above, this would be covered by the first test after the copy

mruberry commented on test/test_torch.py:
```diff
@@ -6433,85 +6442,104 @@ def test_logical_and(self, device):
+                self.assertEqual(Y_expected, X)
+
+    def test_clamp_propagates_nans(self, device):
+        op_list = ('clamp', 'Tensor.clamp', 'clip', 'Tensor.clip')
```
See comments above for how to simplify this

mruberry commented:
This is looking pretty good, @vsimkus. Thanks for taking the time to follow-up. I'm looking forward to get these changes in.

I have a question about removing the vectorized implementations and made some suggestions for simplifying the tests.

vsimkus commented on aten/src/ATen/cpu/vec256/vec256_complex_double.h:
```diff
@@ -406,32 +406,6 @@ Vec256<c10::complex<double>> inline minimum(const Vec256<c10::complex<double>>&
 }
 
-template <>
-Vec256<c10::complex<double>> inline clamp(const Vec256<c10::complex<double>>& a, const Vec256<c10::complex<double>>& min, const Vec256<c10::complex<double>>& max) {
```
I've removed these, since clamp should not support complex as discussed before. 
If we would like to keep these specialized implementations, then the complex implementation in `vec256_base.h` would have to be changed to match the new clamp behavior, which I've removed now as per your suggestion in the previous PR. I.e. any method declarations in `vec256_complex_double.h` (and equivalent specialized implementations) have to be first defined in `vec256_base.h`. 

Note: When you were reviewing I hadn't removed the complex clamp from `vec256_base.h`, I've missed it, but I've now  fixed it and removed it in the edited PR.

mruberry replied:
Oh I see. I wonder if I was misreading. This is just removing the COMPLEX vectorized clamps, not the FLOAT vectorized clamps, right? That sounds brilliant, thank you.

vsimkus replied:
Yes, indeed, this only removes the complex implementation (as indicated by the filename, and the return type of the function) :)

vsimkus commented:
@mruberry I've now updated the PR with the suggested changes, and answered your question above on removing the complex vectorized clamp implementations. Let me know what you think.

mruberry commented on test/test_torch.py:
```diff
@@ -6366,85 +6366,96 @@ def test_logical_and(self, device, dtypes):
+            max_vals = max_vals.cpu().numpy()
+
+        # Use NumPy implementation as reference
+        X_clamped = torch.tensor(np.clip(X.cpu().numpy(), a_min=min_vals, a_max=max_vals))
```
device=device so the tensor is on the same device as X

mruberry reviewed (approved):
Awesome! Thanks @vsimkus. Sorry this review took awhile, we just cut our 1.7 branch and that process was very time consuming. 

Thanks for pointing out we're only removing the complex vectorized clamp impls. 

One last issue before this gets imported is in the docs however. Take a look and let me know your thoughts.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1655,14 +1655,8 @@ def merge_dicts(*dicts):
-        x_i & \text{if } \text{min} \leq x_i \leq \text{max} \\
-        \text{max} & \text{if } x_i > \text{max}
-    \end{cases}
+    y_i = \min(\text{max}, \min(\text{min}, x_i))
```
min(max, min(min, x)?

This is a good lesson in why param names should never be "min" or "max." Two issues here:

- is this correct? I would have expected min(max(x, min), max)?
- how about using different names for the parameters and the functions? This could be done in one of two ways: the functions can be renamed to minimum and maximum or this can use a let expression ("let min_value and max_value be :attr:\`min\` and :attr:\`max\`, respectively, this computes...").

vsimkus replied:
That was a mistake indeed, thanks for spotting it! I've updated this with a "let" expression as you've suggested to redefine the arguments, since on a quick skim it seems like "min" and "max" function names are used more often in the documentation than "minimum" and "maximum". Also, I agree that these argument names aren't great, but changing them would also change function definitions, so I kept the names to what was originally used before.

mruberry replied:
Right. Sorry, didn't mean to suggest you named them.

vsimkus committed:
Standardized clamp implementations across kernels to NumPy-like implementation

vsimkus committed:
Refactored clamp/clip tests, use NumPy as reference implementation

vsimkus committed:
Updated clamp docstring for standardized behavior

vsimkus commented:
@mruberry Thanks again for carefully reviewing the change :) I've updated the documentation and added the small change in torch_test (`device=device`). Also rebased on master, so it has the most recent changes now and can be benchmarked again before merging.

mruberry commented:
Hey @vsimkus, sorry to bother you. Would you mind rebasing this? I'm seeing some odd failures internally and I'm hoping they're in the base revision.

vsimkus commented:
@mruberry Done. I hope the failures go away :)

mruberry commented:
The rebase appears to have solved the issues. Thanks @vsimkus. Looking forward to the next PR in this series. Do you have a good idea for the next step or would you like to discuss some ideas?
